code/modules/mailorder/modr_catalogues.dm

```diff
@@ -0,0 +1,323 @@
+							destination = input(usr, "Enter mail tag without quotes, or SHIP_TO_QM for secure crate-based delivery", src.name, null) as text
+						if(destination && isalive(usr))
+							var/final_bill = src.cartcost //da-na-na na, da-na-na na na
+							var/buy_success = src.shipCart(destination)
+							src.master.display_alert(alert_beep)
+							var/displayMessage = "[bicon(master)] Purchase unsuccessful due to lack of mail-order service to your area."
+							if(buy_success)
+								var/datum/db_record/spender = FindBankAccountByName(src.master.ID_card.registered)
+								if(spender)
+									spender["current_money"] -= final_bill
+								else
+									message_admins("<span class='alert'>[src] tried to charge a card that doesn't exist, yell at kubius</span>")
+								switch(buy_success)
+									if(DELIVERED_TO_MAIL)
+										displayMessage = "[bicon(master)] Thank you for your purchase! Delivery to '[destination]' in progress."
+									if(DELIVERED_TO_QM)
+										displayMessage = "[bicon(master)] Thank you for your purchase! Your items will be sent to the quartermaster's office."
+								src.mode = MODE_LIST
+							src.master.display_message(displayMessage)
+					else
+						src.master.display_alert(alert_beep)
+						var/displayMessage = "[bicon(master)] Purchase failed | [creditCheck]" //period omitted intentionally
+						src.master.display_message(displayMessage)
+
+		if (href_list["clearcart"])
+			if(length(src.cart) > 0)
+				src.cartsize = 0
+				src.cartcost = 0
+				src.cart.Cut()
+
+		if (href_list["add_to_cart"])
+			var/datum/mail_order/F = src.canbuy[href_list["add_to_cart"]]
+			if(istype(F, /datum/mail_order))
+				if(length(F.order_items) + src.cartsize <= 7)
+					src.cartsize += length(F.order_items)
+					src.cartcost += F.cost
+					src.cart += F
+				else
+					var/alert_beep = null
+					if(!src.master.host_program.message_silent)
+						alert_beep = src.master.host_program.message_tone
+					src.master.display_alert(alert_beep)
+					var/displayMessage = "[bicon(master)] Your cart is full! Please place your order or remove an item."
+					src.master.display_message(displayMessage)
+
+		src.master.add_fingerprint(usr)
+		src.master.updateSelfDialog()
+		return
+
+	// arrange for package construction/shipping, then clear cart
+	proc/shipCart(var/destination)
+		var/list/boxstock = list()
+		var/success_style = null //feeds back via return to let purchase process know what to tell the user
+		var/spawn_package_at = null //used for targeting in mail delivery, and just as an integrity check for qm delivery
+		var/fire_package_to = null //ditto
+
+		if(destination == "SHIP_TO_QM")
+			for(var/turf/T in get_area_turfs(/area/supply/spawn_point))
+				spawn_package_at = T
+				break
+			for(var/turf/T in get_area_turfs(/area/supply/delivery_point))
+				fire_package_to = T
+				break
+			if(!spawn_package_at || !fire_package_to)
+			//tried to QM-ship without map support (???), fail gracefully without paying
```

Comment:
the indentation on this comment (and similar comments below) is a bit confusing

Reply:
Cleaned up a bit.